src/Compilers/CSharp/Portable/Syntax/SyntaxEquivalence.cs

```diff
@@ -210,5 +215,46 @@ private static bool AreEquivalentRecursive(GreenNode before, GreenNode after, Fu
                 return true;
             }
         }
+
+        private static bool AreNullableDirectivesEquivalent(GreenNode before, GreenNode after, Func<SyntaxKind, bool> ignoreChildNode)
+        {
+            using var beforeDirectivesEnumerator = ((Green.CSharpSyntaxNode)before).GetDirectives().GetEnumerator();
+            using var afterDirectivesEnumerator = ((Green.CSharpSyntaxNode)after).GetDirectives().GetEnumerator();
+            while (true)
+            {
+                Green.DirectiveTriviaSyntax beforeAnnotation = getNextNullableDirective(beforeDirectivesEnumerator, ignoreChildNode);
+                Green.DirectiveTriviaSyntax afterAnnotation = getNextNullableDirective(afterDirectivesEnumerator, ignoreChildNode);
+
+                if (beforeAnnotation == null || afterAnnotation == null)
```

Comment:
it feels like some of these checks are already done in AreEquivalentRecursive

Reply:
I think you're right, I can just call `AreEquivalentRecursive` and skip the rest of the checks.

---
In reply to: [333772553](https://github.com/dotnet/roslyn/pull/39227#discussion_r333772553) [](ancestors = 333772553)

Reply:
The null checks are the base case, and can't be removed, but the reference equals test can be.

---
In reply to: [333777425](https://github.com/dotnet/roslyn/pull/39227#discussion_r333777425) [](ancestors = 333777425,333772553)